agent/artifact_uploader.go

```diff
@@ -110,87 +110,161 @@ func isDir(path string) bool {
 	return fi.IsDir()
 }
 
-func (a *ArtifactUploader) Collect(ctx context.Context) (artifacts []*api.Artifact, err error) {
+func (a *ArtifactUploader) Collect(ctx context.Context) ([]*api.Artifact, error) {
 	wd, err := os.Getwd()
 	if err != nil {
 		return nil, fmt.Errorf("getting working directory: %w", err)
 	}
 
-	// file paths are deduplicated after resolving globs etc
-	seenPaths := make(map[string]bool)
+	workCh := make(chan string)
+	var wg sync.WaitGroup
 
+	ac := &artifactCollector{
+		ArtifactUploader: a,
+		wd:               wd,
+		seenPaths:        make(map[string]bool),
+	}
+
+	// Start N workers in a new context
+	wctx, cancel := context.WithCancelCause(ctx)
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
is `runtime.NumCPU()` being used here as "an arbitrary smallish number to parallelise by, which will be roughly proportional to the size of the box we're running on"? that's a totally valid use case, i just wanna be sure i'm reading right

Reply:
Yes. It's not a great heuristic though. We could create N goroutines and let the runtime figure it out. The code would be simpler, and unlike a server, the amount of work is fixed ahead of time, so it's not unreasonable. Actually, that idea is growing on me...